### PR TITLE
nvme: ignore NVMe-over-Fabrics devices

### PR DESCRIPTION
**Pull request by ematsumiya**
Ignore NVMe-over-Fabrics devices when probing. Otherwise, logs might be
flooded with error messages:

03:10:53:0251 FuEngine             failed to add device /sys/devices/virtual/nvme-fabrics/ctl/nvme1: failed to find device with subsystems pci, only got nvme,nvme-fabrics

These devices are not supported as they are not on the PCI subsystem.

Signed-off-by: Enzo Matsumiya <ematsumiya@suse.de>

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/nvme/fu-nvme-device.c [hunk 1]
```diff
@@ -267,6 +267,29 @@ fu_nvme_device_dump(const gchar *title, const guint8 *buf, gsize sz)
...
+
+	gdev = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+
+	device_tmp = g_udev_device_get_parent_with_subsystem(gdev, "nvme-fabrics", NULL);
+	if (device_tmp != NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "device is NVMe-over-Fabrics");
+		return TRUE;
```
Comment (superm1) [inline plugins/nvme/fu-nvme-device.c:287]: It's unusual for return true on the failure scenario in this project. Could you instead swap the logic around?

Maybe just match for pci in this new function and if it's anything else fail?

File: plugins/nvme/fu-nvme-device.c [hunk 2]
```diff
@@ -280,6 +303,10 @@ fu_nvme_device_probe(FuDevice *device, GError **error)
 	if (g_strcmp0(fu_device_get_vendor(FU_DEVICE(device)), "Samsung Electronics Co Ltd") == 0)
 		fu_device_set_vendor(FU_DEVICE(device), "Samsung");
 
+	/* ignore NVMe-over-Fabrics devices */
+	if (fu_nvme_device_is_fabrics(device, error))
```
Comment (superm1) [inline plugins/nvme/fu-nvme-device.c:307]: Usually we match the fail here in this project.

**Review by superm1 — CHANGES_REQUESTED** [review]

File: plugins/nvme/fu-nvme-device.c (hunk 1 above)
Comment (ematsumiya) [reply plugins/nvme/fu-nvme-device.c]: Thanks. I force pushed an amend with those changes. Please review.

File: plugins/nvme/fu-nvme-device.c [hunk 3]
```diff
@@ -267,6 +267,33 @@ fu_nvme_device_dump(const gchar *title, const guint8 *buf, gsize sz)
...
+ * %TRUE: device is in PCI subsystem
+ * %FALSE: device is, probably, NVMe-over-Fabrics
+ */
+static gboolean
+fu_nvme_device_is_pci(FuDevice *device, GError **error)
+{
+	g_autoptr(GUdevDevice) device_tmp = NULL;
+	GUdevDevice *gdev;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
```
Comment (superm1) [inline plugins/nvme/fu-nvme-device.c:281]: is this strictly needed?  _Sometimes_ we do like to send `NULL` for the error argument to functions if we really don't care about the details of the outcome.

Comment (hughsie) [reply plugins/nvme/fu-nvme-device.c:281]: In general we don't do these error checks for static functions. I'd say just remove the line then LGTM.

Comment (ematsumiya) [reply plugins/nvme/fu-nvme-device.c:281]: Not strictly needed. While testing my patches, I was getting a segfault on error being dereferenced and I was checking whether it was me not setting error or if it was being derefereced without NULL checks. Turns out it was both. I should send a new patch for the second case though.
For now, I'll remove this line so we can go on.
Thanks.

**Review by hughsie — APPROVED** [review]

**Review by superm1 — APPROVED** [review]

**Pull request merged**